website/blog/2023-4-10-incrementals-pt2.md

```diff
@@ -0,0 +1,538 @@
+- both folders and files are returned in the same request
+- Graph API always returns all returns ancestor folders for an item before the
+  item itself. Folders are returned in the order of the hierarchy (
+  `[root, folder1, ...]`)
+- items in a deleted folder are also returned as deleted
+- `folder3` and its items isn’t returned because nothing in that subtree was
+  changed
+- moving a folder subtree only returns a result for the root of the subtree
+  unless other items in the subtree were individually updated (e.x. no result
+  was returned for `folder5` even though it was moved)
+
+For every returned item, Corso checks if it’s a folder or a file and reacts accordingly.
+
+### Handling folders
+
+Folders have an in-memory representation of a Collection during a backup. This
+helps Corso group items together and allows Corso to express hierarchy changes
+the folder may have participated in.
+
+Every Collection has a current and previous path, representing where the folder
+is in the hierarchy now and where the folder was in the previous backup. The
+current path is sourced from the returned results; it can be generated by
+appending the name of an item to the `parentPath` of the item. Folders that have
+been deleted have their current path set to an empty value.
+
+The previous path is found by looking up the folder’s ID in the `folderMap.json`
+data from the previous backup. Since that map is indexed by ID, it can still
+find folders that have been deleted or folders that have been renamed. Folders
+that weren't in the previous backup (created between when the last backup
+completed and this backup started) have their previous path set to an empty
+value.
+
+Together, the values of the current and previous paths allow Corso to represent
+a set of changes to the hierarchy in an order-independent fashion. That means
+the Collections always represent the same set of changes no matter what order
+other components of Corso see the Collections in. The “state” of a Collection
+can always be determined by comparing the values of the two paths. The table
+below shows the different possible states.
+
+| Previous Path | Current Path | State |
+| --- | --- | --- |
+| any value | empty | deleted folder |
+| empty | any value | new folder |
+| different from current | different from previous | folder moved |
+| same as current | same as previous | no change |
+
+To see where this representation and order-independence really comes in handy,
+consider the case of `folder2`. Between the backups the original `folder2` and
+all it’s items was deleted and a new folder with the same path was created with
+a new item. From the user’s perspective and if potential differences in
+permissions are ignored, there’s not really a difference between the old and new
+`folder2`s. The items they have may be different, but the user generally doesn’t
+inspect the M365 IDs of folders, so the fact that they reside at the same
```

Comment:
There are 10 references to M365 in these blog posts. Please replace with Microsoft 365.

Reply:
each instance replaced